### PR TITLE
added generic template to object-fetching db methods

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

**Comment by DmitryTronin** [comment]
There are conflicts between master & your branch. Could you please run through them & push again?

**Comment by staabm** [comment]
Sure

**Comment by isfedorov** [comment]
@staabm unfortunately, there is still conflict with the master branch that prevents changes from merging.

File: PDO/PDO.php [hunk 1]
```diff
@@ -1679,16 +1679,18 @@ public function fetchAll(
...
      * @link https://php.net/manual/en/pdostatement.fetchobject.php
-     * @param string $class [optional] <p>
+     * @param class-string<T> $class [optional] <p>
      * Name of the created class.
      * </p>
      * @param array $constructorArgs [optional] <p>
      * Elements of this array are passed to the constructor.
      * </p>
-     * @return mixed an instance of the required class with property names that
+     * @return T an instance of the required class with property names that
```
Comment (staabm) [inline PDO/PDO.php:1693]:
```suggestion
     * @return T|stdClass|null an instance of the required class with property names that
```

File: mysql/mysql.php [hunk 2]
```diff
@@ -444,18 +444,20 @@ function mysql_fetch_array($result, $result_type = MYSQL_BOTH) {}
...
+ * @param class-string<T> $class_name [optional] <p>
  * The name of the class to instantiate, set the properties of and return.
  * If not specified, a <b>stdClass</b> object is returned.
  * </p>
  * @param array $params [optional] <p>
  * An optional array of parameters to pass to the constructor
  * for <i>class_name</i> objects.
  * </p>
- * @return stdClass|object an object with string properties that correspond to the
+ * @return T|object an object with string properties that correspond to the
```
Comment (staabm) [inline mysql/mysql.php:460]:
```suggestion
 * @return T|stdClass an object with string properties that correspond to the
```

File: mysqli/mysqli.php [hunk 3]
```diff
@@ -1229,17 +1229,19 @@ public function fetch_array(int $mode = MYSQLI_BOTH): array|false|null {}
...
+     * @param class-string<T> $class [optional] <p>
      * The name of the class to instantiate, set the properties of and return.
      * If not specified, a <b>stdClass</b> object is returned.
      * </p>
      * @param null|array $constructor_args [optional] <p>
      * An optional array of parameters to pass to the constructor
      * for <i>class_name</i> objects.
      * </p>
-     * @return object|false|null an object representing the fetched row, where each property
+     * @return T|object|null an object representing the fetched row, where each property
```
Comment (staabm) [inline mysqli/mysqli.php:1244]:
```suggestion
     * @return T|stdClass|false|null an object representing the fetched row, where each property
```

File: mysqli/mysqli.php [hunk 4]
```diff
@@ -1881,13 +1883,15 @@ function mysqli_fetch_array(mysqli_result $result, int $mode = MYSQLI_BOTH): arr
...
+ * 
  * Fetch the next row of a result set as an object
  * @link https://php.net/manual/en/mysqli-result.fetch-object.php
  * @param mysqli_result $result A mysqli_result object returned by mysqli_query(),
  * mysqli_store_result(), mysqli_use_result() or mysqli_stmt_get_result().
- * @param string $class [optional] The name of the class to instantiate, set the properties of and return. If not specified, a stdClass object is returned.
+ * @param class-string<T> $class [optional] The name of the class to instantiate, set the properties of and return. If not specified, a stdClass object is returned.
  * @param array $constructor_args [optional] An optional array of parameters to pass to the constructor for class_name objects.
- * @return object|null|false an object representing the fetched row,
+ * @return T|null|false an object representing the fetched row,
```
Comment (staabm) [inline mysqli/mysqli.php:1894]:
```suggestion
 * @return T|stdClass|null|false an object representing the fetched row,
```

**Comment by staabm** [comment]
rebased

**Pull request merged**